### PR TITLE
Create a Location Hierarchy API

### PR DESCRIPTION
**Pull request by rehammuzzamil**
https://github.com/opensrp/fhircore/issues/188

### CODE REVIEW

**Review by pld — COMMENTED** [review]
can we write test for this?

**Comment by rehammuzzamil** [comment]
> can we write test for this?

Sure, will add unit tests for this repo.

**Review by pld — APPROVED** [review]
lgtm!

**Review by pld — APPROVED** [review]

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            <scope>provided</scope>
+        </dependency>
+
+        <!-- This dependency includes the core HAPI-FHIR classes -->
+        <dependency>
+            <groupId>ca.uhn.hapi.fhir</groupId>
+            <artifactId>hapi-fhir-base</artifactId>
+            <version>5.4.0</version>
+        </dependency>
+
```
Comment (dubdabasoduba) [inline pom.xml:60]: Please add the publishing settings


```
<distributionManagement>
		<snapshotRepository>
			<id>nexus-snapshots</id>
			<name>Nexus Snapshots Repository</name>
			<url>https://oss.sonatype.org/content/repositories/snapshots</url>
			<uniqueVersion>false</uniqueVersion>
		</snapshotRepository>
		<repository>
			<id>nexus-releases</id>
			<url>https://oss.sonatype.org/service/local/staging/deploy/maven2</url>
		</repository>
	</distributionManagement>
```

File: src/main/java/org/smartregister/extension/model/LocationHierarchy.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	@Child(name = "LocationHierarchyTree", type = { LocationHierarchyTree.class })
+	@Description(
+			shortDefinition = "Complete Location Hierarchy Tree",
+			formalDefinition = "Consists of Location Hierarchy Tree and Parent Child Identifiers List"
+	)
+	private LocationHierarchyTree locationHierarchyTree;
+
+	@Override
+	public Location copy() {
+		Location dst = new Location();
```
Comment (dubdabasoduba) [inline src/main/java/org/smartregister/extension/model/LocationHierarchy.java]: Use a better name other that `dst`

File: src/main/java/org/smartregister/extension/model/LocationHierarchy.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		Bundle bundle = new Bundle();
+		List<Bundle.BundleEntryComponent> theEntry = new ArrayList<>();
```
Comment (dubdabasoduba) [inline src/main/java/org/smartregister/extension/model/LocationHierarchy.java:40]: Please name the variables according to what the hold or do

File: src/main/java/org/smartregister/extension/model/Tree.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+				}
+			}
+		}
+
+		if (kids == null) {
+			kids = new ArrayList<>();
+		}
+		StringType idStringType = new StringType();
+		String idString = id;
+		if (idString.contains("/_")) {
```
Comment (dubdabasoduba) [inline src/main/java/org/smartregister/extension/model/Tree.java]: Move this to a constants file

File: src/main/java/org/smartregister/extension/model/Tree.java [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+			idString = idString.substring(0, idString.indexOf("/_"));
```
Comment (dubdabasoduba) [inline src/main/java/org/smartregister/extension/model/Tree.java]: Move this to a constants file

File: src/main/java/org/smartregister/extension/model/Tree.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+		}
+		idStringType.setValue(idString);
+
+		StringType parentStringType = new StringType();
+		parentStringType.setValue(parent);
+		kids.add(idStringType);
+		Boolean setParentChildMap = false;
+		for (int i = 0; i < parentChildren.size(); i++) {
+			if (parentChildren.get(i) != null && parentChildren.get(i).getIdentifier() != null
```
Comment (dubdabasoduba) [inline src/main/java/org/smartregister/extension/model/Tree.java:78]: When checking Strings use `StringUtils.isNotBlank`

File: src/main/java/org/smartregister/extension/model/Tree.java [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+		if (listOfNodes == null) {
+			listOfNodes = new SingleTreeNode();
+		}
+
+		// if node exists we should break since user should write optimized code and also tree can not have duplicates
+		if (hasNode(id)) {
+			throw new IllegalArgumentException("Node with ID " + id + " already exists in tree");
+		}
+
+		TreeNode n = makeNode(id, label, node, parentId);
```
Comment (dubdabasoduba) [inline src/main/java/org/smartregister/extension/model/Tree.java]: Rename this to a better variable name

File: src/main/java/org/smartregister/extension/model/Tree.java [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+		if (parentId != null) {
+			addToParentChildRelation(parentId, id);
+
+			TreeNode p = getNode(parentId);
```
Comment (dubdabasoduba) [inline src/main/java/org/smartregister/extension/model/Tree.java]: Rename to a better variable name

File: src/main/java/org/smartregister/extension/model/Tree.java [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+			//if parent exists add to it otherwise add as root for now
+			if (p != null) {
+				p.addChild(n);
+			} else {
+				// if no parent exists add it as root node
+				String idString = (String) id;
+				if (idString.contains("/_")) {
```
Comment (dubdabasoduba) [inline src/main/java/org/smartregister/extension/model/Tree.java]: Move the `"/_"` to constants

File: src/main/java/org/smartregister/extension/model/Tree.java [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+					idString = idString.substring(0, idString.indexOf("/_"));
```
Comment (dubdabasoduba) [inline src/main/java/org/smartregister/extension/model/Tree.java]: Move the `"/_"` to constants

File: src/main/java/org/smartregister/extension/model/Tree.java [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+				treeNodeId.setValue(idString);
+				singleTreeNode.setTreeNodeId(treeNodeId);
+				singleTreeNode.setTreeNode(n);
+				listOfNodes = singleTreeNode;
+			}
+		} else {
+			// if no parent add it as root node
+			String idString = id;
+			if (idString.contains("/_")) {
+				idString = idString.substring(0, idString.indexOf("/_"));
```
Comment (dubdabasoduba) [inline src/main/java/org/smartregister/extension/model/Tree.java]: Move the `"/_"` to constants

File: src/main/java/org/smartregister/extension/model/Tree.java [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+			StringType treeNodeId = new StringType();
+			treeNodeId.setValue(idString);
+			singleTreeNode.setTreeNodeId(treeNodeId);
+			singleTreeNode.setTreeNode(n);
+			listOfNodes = singleTreeNode;
+		}
+	}
+
+	private TreeNode makeNode(String id, String label, Location node, String parentId) {
+		TreeNode n = getNode(id);
```
Comment (dubdabasoduba) [inline src/main/java/org/smartregister/extension/model/Tree.java]: Rename the variables

File: src/main/java/org/smartregister/extension/rest/LocationHierarchyResourceProvider.java [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+	private List<Location> getLocationHierarchy(Long id, IBaseResource parentLocation) {
+		return descendants(id, parentLocation);
+	}
+
+	public List<Location> descendants(Long id, IBaseResource parentLocation) {
+
+		SearchParameterMap paramMap = new SearchParameterMap();
+		ReferenceAndListParam thePartOf = new ReferenceAndListParam();
+		ReferenceParam partOf = new ReferenceParam();
+		partOf.setValue("Location/" + id);
```
Comment (dubdabasoduba) [inline src/main/java/org/smartregister/extension/rest/LocationHierarchyResourceProvider.java]: Please move all hardcoded strings to a constants file

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 1 above)
Comment (rehammuzzamil) [reply pom.xml:60]: Added

File: src/main/java/org/smartregister/extension/model/LocationHierarchy.java (hunk 2 above)
Comment (rehammuzzamil) [reply src/main/java/org/smartregister/extension/model/LocationHierarchy.java]: Done

File: src/main/java/org/smartregister/extension/model/Tree.java (hunk 4 above)
Comment (rehammuzzamil) [reply src/main/java/org/smartregister/extension/model/Tree.java]: Created constants file.

File: src/main/java/org/smartregister/extension/model/Tree.java (hunk 12 above)
Comment (rehammuzzamil) [reply src/main/java/org/smartregister/extension/model/Tree.java]: Done

File: src/main/java/org/smartregister/extension/model/Tree.java (hunk 6 above)
Comment (rehammuzzamil) [reply src/main/java/org/smartregister/extension/model/Tree.java:78]: Done

File: src/main/java/org/smartregister/extension/model/Tree.java (hunk 7 above)
Comment (rehammuzzamil) [reply src/main/java/org/smartregister/extension/model/Tree.java]: Done

File: src/main/java/org/smartregister/extension/model/Tree.java (hunk 8 above)
Comment (rehammuzzamil) [reply src/main/java/org/smartregister/extension/model/Tree.java]: Done

**Review by dubdabasoduba — APPROVED** [review]

**Pull request merged**